### PR TITLE
LOGIN

### PR DESCRIPTION
**Pull request by AbdallahAmmar96**

### CODE REVIEW

File: server/package.json [hunk 1]
```diff
@@ -12,9 +12,12 @@
   },
   "license": "ISC",
   "dependencies": {
+    "cookie-parser": "^1.4.4",
+    "cors": "^2.8.5",
```
Comment (kannanhassouna) [inline server/package.json]:
```suggestion
    "cors": "^2.8.5",
```
remove cors

File: server/helpers/validation-schema.js [hunk 2]
```diff
@@ -25,3 +25,8 @@ exports.applyCourseSchema = joi.object().keys({
   datesId: joi.number().integer(),
   studentid: joi.number().integer(),
 });
+
+exports.loginSchema = joi.object().keys({
+  id: joi.number().integer(),
```
Comment (kannanhassouna) [inline server/helpers/validation-schema.js]: do you thing that's a good schema

File: server/database/config/db_build.sql [hunk 3]
```diff
@@ -15,7 +15,7 @@ CREATE TABLE student (
 );
 
 CREATE TABLE staff (
-	id SERIAL PRIMARY KEY,
+	id INT,
```
Comment (kannanhassouna) [inline server/database/config/db_build.sql:18]:
```suggestion
	id INT primary key,
```

File: server/controllers/index.js [hunk 4]
```diff
@@ -2,15 +2,27 @@ const router = require('express').Router();
 
 const student = require('./student');
 const courses = require('./courses');
-// const staff = require('./staff');
+const login = require('./auth/login');
+const logout = require('./logout');
+const auth = require('../middlewares/authentication');
+const authFrontEnd = require('../controllers/auth/auth');
```
Comment (kannanhassouna) [inline server/controllers/index.js:8]:
```suggestion
const checkCookie = require('../controllers/auth/checkCookie');
```

File: server/controllers/auth/auth.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        res.send({
+          isAuth: false,
+        });
+      });
+  } else {
+    res.send({
+      isAuth: false,
+    });
+  }
+};
```
Comment (kannanhassouna) [inline server/controllers/auth/auth.js]: just a new line

File: server/controllers/auth/login.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+module.exports = (req, res) => {
+  const { id, password } = req.body;
+  const { error } = joi.validate({ id, password }, loginSchema);
+  if (error) {
+    res.status(400).send({
+      data: null,
+      error: error.details[0].message,
+    });
+  } else if (id.length === 5) {
```
Comment (kannanhassouna) [inline server/controllers/auth/login.js]: i think that should be in validation .
```suggestion
  } else if (id.length === 5) {
```

File: server/controllers/auth/login.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      .then((result) => {
+        if (result.rowCount) {
+          const payLoad = {
+            id: result.rows[0].id,
+            role: 'student',
+          };
+          const jwt = createCookie(payLoad);
+          res.cookie('jwt', jwt, {
+            maxAge: 7200000,
+            httpOnly: true,
```
Comment (kannanhassouna) [inline server/controllers/auth/login.js]: you should insert secure property .
```suggestion
            secure: true,
```

File: server/controllers/auth/login.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+          });
+          res.send({
+            error: null,
+            data: result.rows,
```
Comment (kannanhassouna) [inline server/controllers/auth/login.js]: you can't send the password to user in the response because anyone can intercept the response

File: server/controllers/auth/login.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+          });
+        }
+      })
+      .catch(() => {
+        res.status(500).send({
+          error: 'Internal Server Error',
+          data: null,
+        });
+      });
+  } else if (id.length === 6) {
```
Comment (kannanhassouna) [inline server/controllers/auth/login.js]: you should rely on something else. (let me thing about that )

File: Client/public/index.html [hunk 10]
```diff
@@ -15,7 +15,7 @@
       manifest.json provides metadata used when your web app is installed on a
       user's mobile device or desktop. See https://developers.google.com/web/fundamentals/web-app-manifest/
     -->
-    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
+    <!-- <link rel="manifest" href="%PUBLIC_URL%/manifest.json" /> -->
```
Comment (kannanhassouna) [inline Client/public/index.html:18]: remove this file from fP.R

File: Client/src/components/App.js [hunk 11]
```diff
@@ -1,13 +1,55 @@
...
-      {/* <Descktop /> */}
-    </div>
-  )
+import Descktop from './Layout/Desktop';
+import axios from 'axios';
+
+export default class App extends Component {
+  state = {
+    auth: false,
+    isLogin: false,
```
Comment (kannanhassouna) [inline Client/src/components/App.js:13]: what's the different between auth and isLogin ?

File: Client/src/components/App.js [hunk 12]
```diff
@@ -1,13 +1,55 @@
...
-      <Mobile />
-      {/* <Descktop /> */}
-    </div>
-  )
+import Descktop from './Layout/Desktop';
+import axios from 'axios';
+
+export default class App extends Component {
+  state = {
+    auth: false,
```
Comment (tomduggan85) [inline Client/src/components/App.js:12]: @AbdallahAmmar96 I think changing the name of `auth` to `hasLoaded` would make the code a little easier to understand

File: Client/src/components/App.js [hunk 13]
```diff
@@ -1,13 +1,55 @@
...
+
+  handleLogin = (id) => {
+    this.setState({ isLogin: true, id })
+  }
+
+  render() {
+    const { auth, isLogin, role, id } = this.state;
+    return (auth ?
+      <Router>
+        {isLogin ? (id.toString().length === 5 ?
```
Comment (tomduggan85) [inline Client/src/components/App.js]: @AbdallahAmmar96 It's a little hard to follow all of the nested ternary operators in this render - I have a few ideas on how to rearrange this and make it easier to follow, if you'd like:
1) Move the `<h1>Loading</h1>` to the top, to short-circuit out of the render function early:
```
render() {
  const { auth, isLogin, role, id } = this.state;
  if (!auth) {
    return <h1>Loading..</h1>;
  }

  return (
    <Router>
    .....
```

2. Move the id.toString() check into a variable, and then conditionally include each route based on it:
```
const isStudent = id.toString().length === 5
...
{isStudent && <Route path="/student" component={Mobile} />}
{!isStudent && <Route path="/staff" component={Desktop} />}
...
```
Also, if the string length check could be changed into something more exact, like id === "(something goes here)", that would help.

File: server/test/Endpoints/router-test.js [hunk 14]
```diff
@@ -3,11 +3,13 @@ const request = require('supertest');
 const app = require('../../app');
 const dbBuild = require('../../database/config/db_build');
 
+
 test('Get all student information from /api/v1/student/', (t) => {
   dbBuild()
     .then(() => {
       request(app)
         .get('/api/v1/student/')
+        .set('Cookie', ['jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTIzNDUsInJvbGUiOiJzdHVkZW50IiwiaWF0IjoxNTU2NjA2MjQ4fQ.jxmj_2S-sGc7XKoZPkh0JDDTR1AoWJvNggokoC54QP4'])
```
Comment (tomduggan85) [inline server/test/Endpoints/router-test.js:12]: @AbdallahAmmar96 You could move this cookie value into a const variable, and then if you ever have to change it in the future, you only have to change one line instead of many lines.

File: Client/src/components/contanier/Auth/index.js [hunk 15]
```diff
@@ -1,27 +1,136 @@
...
+      <
+      img src = {
+        logo
+      }
+      alt = "amideast logo"
+      className = 'login-logo' / >
+      <
+      div className = 'mobile-container' >
+      <
+      LabeledInput LabeledInputClassName = 'container'
```
Comment (tomduggan85) [inline Client/src/components/contanier/Auth/index.js]: @AbdallahAmmar96 If you have time, I think this code would be easier to read if you put all of the opening / closing brackets onto the same line as the tag and indented child tags, so for example change it from this:
```
<
div className='login>
<
img src=...
```
to this
```
<div className='login>
  <img src=...
```

**Comment by Fatmasiam** [comment]
solve Travis ic problem

File: Client/src/components/contanier/Auth/index.js [hunk 16]
```diff
@@ -1,27 +1,136 @@
...
 
+  state = {
+    id: '',
+    password: '',
+    loginError: '',
+    role: '',
+  };
+
+  handleChange = ({
+    target: {
```
Comment (Fatmasiam) [inline Client/src/components/contanier/Auth/index.js:21]: try to write this in one line 
like :   ` handleChange = ({ target: { name, value } }) => this.setState({ [name]: value.trim() });`
but I have sth I want to inquire about it if I am writing password like 12[space-space-space] 
then my password will be like this `12` so if I trying to log in every time I can't log in and don't understand where is the problem though I am ensuring that I writing the correct password

File: Client/src/components/contanier/Auth/index.js [hunk 17]
```diff
@@ -1,27 +1,136 @@
...
+      }
+      />  <
+      h2 > {
+        this.state.loginError
+      } < /h2> < /
+      div >
     )
   }
-}
+}
```
Comment (Fatmasiam) [inline Client/src/components/contanier/Auth/index.js]: and a new line here

File: Client/src/components/Layout/SideBar/NavElement/index.js [hunk 18]
```diff
@@ -1,12 +1,27 @@
...
+      ?
+      (
+        <NavLink to={link} onClick={handleLinkClick} className='SidBar-link' >
+          <img src={icon} alt="icon" className='SideBar-img' />
+          <h3 className='SideBar-text'>{text}</h3>
+        </NavLink>)
+      :
+      // (
+        // <NavLink to={'/login'} onClick={handleLinkClick} className='SidBar-link' >
+        //   <img src={icon} alt="icon" className='SideBar-img' />
```
Comment (Fatmasiam) [inline Client/src/components/Layout/SideBar/NavElement/index.js]: Command code !!!

File: Client/src/components/contanier/Auth/index.js [hunk 19]
```diff
@@ -1,27 +1,136 @@
...
+      id,
+      password
+    } = this.state;
+    if (!id || !password) {
+      this.setState({
+        loginError: 'All Fields Requier'
+      })
+    } else if (id.length === 0) {
+      this.setState({
+        loginError: 'ID At least Five Numbers'
```
Comment (Fatmasiam) [inline Client/src/components/contanier/Auth/index.js:43]: are u ensure that your error statement suitable with the condition? 
the condition ` id.length === 0` and the statement  `ID At least Five Numbers` I think there is a conflict between them.

Comment (Fatmasiam) [reply Client/src/components/contanier/Auth/index.js:43]: and write the error statement like: `ID at least five numbers`

File: server/controllers/auth/auth.js (hunk 5 above)
Comment (Fatmasiam) [inline server/controllers/auth/auth.js]: new line

File: Client/src/components/contanier/student/Apply/index.js [hunk 20]
```diff
@@ -15,7 +16,15 @@ export default class Apply extends Component {
...
   }
+
+  handleClick (e) {
+    e.preventDefault();
+    axios
+    .post()
+  }
   render() {
+    console.log(this.props)
+    console.log('99999999999999999999999999')
```
Comment (Fatmasiam) [inline Client/src/components/contanier/student/Apply/index.js:27]: remove console.log

File: Client/src/components/contanier/student/Coures/index.js [hunk 21]
```diff
@@ -1,28 +1,46 @@
...
-    e.preventDefault()
+  state = {
+    data: null,
+  }
+
+  componentDidMount() {
+    if (this.props.level) {
+      // ${this.props.level + 1}
+      axios
+        .get(`/api/v1/course/level/3`)
```
Comment (Fatmasiam) [inline Client/src/components/contanier/student/Coures/index.js:19]: I think that you should pass the level number here not static data.

File: Client/src/components/contanier/student/Coures/index.js [hunk 22]
```diff
@@ -1,28 +1,46 @@
...
 
-  handleBtnClick = (e) => {
-    e.preventDefault()
+  state = {
+    data: null,
+  }
+
+  componentDidMount() {
+    if (this.props.level) {
+      // ${this.props.level + 1}
```
Comment (Fatmasiam) [inline Client/src/components/contanier/student/Coures/index.js:17]: here we already stored the target level so u shouldn't increment the level number with 1

File: Client/src/components/contanier/Auth/index.js [hunk 23]
```diff
@@ -1,27 +1,117 @@
...
+    } else {
+      this.setState({
+        loginError: ''
+      });
+      axios
+        .post('/api/v1/login', {
+          id,
+          password
+        })
+        .then(res => {
```
Comment (Fatmasiam) [inline Client/src/components/contanier/Auth/index.js:58]:
```suggestion
  .then(res => {
     const { error } = res.data;
     const { data } = res.data; 
```
**just to be more readable**

File: Client/src/components/contanier/student/Coures/index.js [hunk 24]
```diff
@@ -1,28 +1,46 @@
...
   
   render() {
+    
+    const { data } = this.state;
     return (
+      !data ? <h2>Loading ...</h2> :
       <div className='courses'>
         <h1 className='courses-title'>Available Courses</h1>
         {data.map( item => {
+          console.log(item.id)
```
Comment (Fatmasiam) [inline Client/src/components/contanier/student/Coures/index.js:34]:
```suggestion
```

File: Client/src/components/contanier/student/StudentCourses/index.js [hunk 25]
```diff
@@ -1,13 +1,28 @@
...
+    data: null,
+  }
+
+  componentDidMount() {
+    axios
+      .get(`/api/v1/course/student/${this.props.id}`)
+      .then((res) => {
+        this.setState({data: res.data.data})
+      })
+      .catch(console.log)
```
Comment (Fatmasiam) [inline Client/src/components/contanier/student/StudentCourses/index.js:19]: console log again !!! 
try to handle the error

File: Client/src/components/contanier/Auth/index.js [hunk 26]
```diff
@@ -1,27 +1,117 @@
...
+            }
+          } else {
+            this.setState({
+              loginError: 'Check your ID or Password'
+            })
+          }
+        })
+        .catch((e) => {
+          this.setState({
+            loginError: 'Try Againe'
```
Comment (Fatmasiam) [inline Client/src/components/contanier/Auth/index.js:81]: **typo** 
```suggestion
            loginError: 'Try again'
```

File: Client/src/components/App.js [hunk 27]
```diff
@@ -1,13 +1,55 @@
...
+  render() {
+    const {auth, isLogin, id} = this.state;
+    const isStudent = id.toString().length === 5;
+    
+    return(
+      !auth ? <h2>Loading ...</h2> : 
+      <Router>
+      {isLogin ?  
+        <Switch>
+          {isStudent && <Route path='/student' component={ (props) => <Mobile id={id} {...props} />} />}
```
Comment (Fatmasiam) [inline Client/src/components/App.js]:
```suggestion
          {isStudent ?
           <Route path='/student' component={ (props) => <Mobile id={id} {...props} />} />
           :
          <Route path='/staff' component={Desktop} />
           }
```

File: Client/src/components/Layout/Mobile/index.js [hunk 28]
```diff
@@ -1,50 +1,59 @@
...
+    const { id } = this.props;
+    const { menuOpen } = this.state;
     const { handleMenuClick, handleLinkClick } = this;
     return (
       <BrowserRouter className='mobile-container'>
-        {login ? (
-          <>
             <Header handleMenuClick={handleMenuClick} menuOpen={menuOpen} />
-            <SideBar handleLinkClick={handleLinkClick} menuOpen={menuOpen} />
+            <SideBar handleLinkClick={handleLinkClick} menuOpen={menuOpen}  history={this.props.history} />
```
Comment (naremanhilles) [inline Client/src/components/Layout/Mobile/index.js:46]: can you explain props.history,  please?
